File: control-panel/shared/src/main/java/com/github/dynamicextensionsalfresco/controlpanel/template/TemplateBundle.java
package com.github.dynamicextensionsalfresco.controlpanel.template;

import static java.util.Collections.emptyList;

import com.github.dynamicextensionsalfresco.osgi.BundleHelpers;
import com.github.dynamicextensionsalfresco.controlpanel.BundleHelper.BundleHelperCompanion;
import com.springsource.util.osgi.manifest.BundleManifest;
import com.springsource.util.osgi.manifest.BundleManifestFactory;
import com.springsource.util.osgi.manifest.ExportedPackage;
import com.springsource.util.osgi.manifest.ImportedPackage;
import org.osgi.framework.Bundle;
import org.osgi.framework.Constants;
import org.osgi.framework.ServiceReference;
import java.text.SimpleDateFormat;
import java.util.*;
import org.osgi.framework.Version;

/**
 * Adapts an [Bundle] for display in a Freemarker template.

 * @author Laurens Fridael
 */
public class TemplateBundle implements Comparable<TemplateBundle> {
    private int FRAMEWORK_BUNDLE_ID = 0;

    public TemplateBundle(Bundle bundle) {
        this(bundle, null);
    }
    public TemplateBundle(Bundle bundle, List<ServiceReference<Object>> services) {
        this.bundle = bundle;
        init(services);
    }

    private Bundle bundle;
    private List<TemplateServiceReference> services = null;

    public List<TemplateServiceReference> getServices() {
        return this.services;
    }

    private void init(List<ServiceReference<Object>> initServices){
        if(initServices == null) {
            this.services = emptyList();
            return;
        }
        this.services = Arrays.asList(
                            initServices.stream()
                                    .map(TemplateServiceReference::new)
                                    .sorted()
                                    .toArray(TemplateServiceReference[]::new)
                        );
    }

    public Long getBundleId(){
        return bundle != null ? bundle.getBundleId() : 0;
    }

    public String getSymbolicName() {
        return bundle != null && bundle.getSymbolicName() != null
                ? bundle.getSymbolicName()
                : "non OSGi jar file";
    }

    public String getName() {
        return getHeader(Constants.BUNDLE_NAME) != null
                ? getHeader(Constants.BUNDLE_NAME)
                : getSymbolicName();
    }

    public String getDescription() {
        return getHeader(Constants.BUNDLE_DESCRIPTION);
    }

    private String getHeader(String header) {
        return bundle != null && bundle.getHeaders() != null
                ? bundle.getHeaders().get(header)
                : null;
    }

    public Boolean getDynamicExtension() { return BundleHelperCompanion.isDynamicExtension(bundle);}

    public Boolean getFragmentBundle() { return getHeader(Constants.FRAGMENT_HOST) != null; }

    public String getLocation() { return bundle != null ? bundle.getLocation() : ""; }

    public String getLastModified() {
        if(bundle == null) {
            return null;
        }

        long lastModified = bundle.getLastModified();
        if (lastModified > 0) {
            return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z").format(new Date(lastModified));
        } else {
            return null;
        }
    }

    public String getVersion() { return bundle != null ? bundle.getVersion().toString() : ""; }

    public String getStore() {
        String defaultStore = "n/a";
        if(bundle == null){
            return defaultStore;
        }

        if (bundle.getLocation().startsWith("file:")) {
            return "filesystem";
        } else if (bundle.getLocation().startsWith("/")) {
            return "repository";
        } else {
            return defaultStore;
        }
    }

    public String getStatus() {
        if(bundle == null){
            return BundleHelpers.getBundleStateDescription(0);
        }
        return BundleHelpers.getBundleStateDescription(bundle.getState());
    }

    public String getExportPackage() { return getHeader(Constants.EXPORT_PACKAGE); }

    public String getDocumentationUrl() { return getHeader(Constants.BUNDLE_DOCURL); }

    public Boolean getDeleteable() { return this.getLocation().startsWith("/Company Home"); }

    public List<TemplateImportedPackage> getImportedPackages() {
        ArrayList<TemplateImportedPackage> packages = new ArrayList<TemplateImportedPackage>();
        BundleManifest manifest = getManifest();
        if(manifest == null) {
            return packages;
        }

        for (ImportedPackage importedPackage : manifest.getImportPackage().getImportedPackages()) {
            TemplateImportedPackage bundlePackage = new TemplateImportedPackage();
            bundlePackage.setName(importedPackage.getPackageName());
            Version ceiling = importedPackage.getVersion().getCeiling();
            if (ceiling != null) {
                bundlePackage.setMaxVersion(ceiling.toString());
            }
            Version floor = importedPackage.getVersion().getFloor();
            if (floor != null) {
                bundlePackage.setMinVersion(floor.toString());
            }
            packages.add(bundlePackage);
        }
        return packages;
    }

    public List<ExportedPackage> getExportedPackages() {
        return getManifest() != null
                ? getManifest().getExportPackage().getExportedPackages()
                : new ArrayList<ExportedPackage>();
    }

    private BundleManifest manifestCache = null;
    public BundleManifest getManifest() {
        if(bundle == null){
            return null;
        }

        if(manifestCache == null) {
            manifestCache = BundleManifestFactory.createBundleManifest(bundle.getHeaders());
        }

        return manifestCache;
    }


    @Override
    public int compareTo(TemplateBundle other) {
        if(other == null) {
            return -1;
        }

        if (this.getBundleId().equals((long) FRAMEWORK_BUNDLE_ID)) {
            return Integer.MIN_VALUE;
        } else if (other.getBundleId().equals((long) FRAMEWORK_BUNDLE_ID)) {
            return Integer.MAX_VALUE;
        }

        int compare = this.getName().compareToIgnoreCase(other.getName());
        if (compare == 0) {
            return other.getVersion().compareTo(this.getVersion());
        }

        return compare;
    }
}

File: control-panel/src/main/java/com/github/dynamicextensionsalfresco/controlpanel/template/TemplateBundle.kt
package com.github.dynamicextensionsalfresco.controlpanel.template

import com.github.dynamicextensionsalfresco.controlpanel.BundleHelper
import com.github.dynamicextensionsalfresco.osgi.stateDescription
import com.springsource.util.osgi.manifest.BundleManifestFactory
import com.springsource.util.osgi.manifest.ExportedPackage
import org.osgi.framework.Bundle
import org.osgi.framework.Constants
import org.osgi.framework.ServiceReference
import java.text.SimpleDateFormat
import java.util.*

/**
 * Adapts an [Bundle] for display in a Freemarker template.

 * @author Laurens Fridael
 */
class TemplateBundle(private val bundle: Bundle, services: List<ServiceReference<Any?>>?) : Comparable<TemplateBundle> {
    private val FRAMEWORK_BUNDLE_ID = 0

    constructor(bundle: Bundle) : this(bundle, null)

    val services: List<TemplateServiceReference>

    init {
        this.services = services
                    ?.map { TemplateServiceReference(it) }
                    ?.sorted()
                    ?: emptyList()
    }

    val bundleId: Long
        get() = bundle.bundleId

    val symbolicName: String
        get() = bundle.symbolicName ?: "non OSGi jar file"

    val name: String
        get() = bundle.headers.get(Constants.BUNDLE_NAME)?.toString() ?: symbolicName

    val description: String?
        get() = bundle.headers.get(Constants.BUNDLE_DESCRIPTION)?.toString()

    val isDynamicExtension: Boolean
        get() = BundleHelper.isDynamicExtension(bundle)

    val isFragmentBundle: Boolean
        get() = bundle.headers.get(Constants.FRAGMENT_HOST) != null

    val location: String
        get() = bundle.location

    val lastModified: String?
        get() {
            val lastModified = bundle.lastModified
            if (lastModified > 0) {
                return SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z").format(Date(lastModified))
            } else {
                return null
            }
        }

    val version: String
        get() = bundle.version.toString()

    val store: String
        get() {
            if (bundle.location.startsWith("file:")) {
                return "filesystem"
            } else if (bundle.location.startsWith("/")) {
                return "repository"
            } else {
                return "n/a"
            }
        }

    val status: String?
        get() {
            return bundle.stateDescription
        }

    val exportPackage: String?
        get() = bundle.headers.get(Constants.EXPORT_PACKAGE)

    val documentationUrl: String?
        get() = bundle.headers.get(Constants.BUNDLE_DOCURL)

    val isDeleteable: Boolean
        get() = location.startsWith("/Company Home")

    val importedPackages: List<TemplateImportedPackage>
        get() {
            val packages = ArrayList<TemplateImportedPackage>()
            for (importedPackage in manifest.importPackage.importedPackages) {
                val bundlePackage = TemplateImportedPackage()
                bundlePackage.name = importedPackage.packageName
                val ceiling = importedPackage.version.ceiling
                if (ceiling != null) {
                    bundlePackage.maxVersion = ceiling.toString()
                }
                val floor = importedPackage.version.floor
                if (floor != null) {
                    bundlePackage.minVersion = floor.toString()
                }
                packages.add(bundlePackage)
            }
            return packages
        }

    val exportedPackages: List<ExportedPackage>
        get() = manifest.exportPackage.exportedPackages

    val manifest by lazy {
        BundleManifestFactory.createBundleManifest(bundle.headers)
    }


    override fun compareTo(other: TemplateBundle): Int {
        if (this.bundleId == FRAMEWORK_BUNDLE_ID.toLong()) {
            return Integer.MIN_VALUE
        } else if (other.bundleId == 0L) {
            return Integer.MAX_VALUE
        }

        val compare = this.name.compareTo(other.name, ignoreCase = true)
        if (compare == 0) {
            return this.version.compareTo(other.version)

        }

        return compare
    }
}
